lib/wasi/src/state/builder.rs

```diff
@@ -792,40 +793,197 @@ impl WasiEnvBuilder {
 
-        env.data(store).thread.set_status_running();
-        let mut res = crate::run_wasi_func_start(start, store);
+        let result = crate::run_wasi_func_start(start, store);
+        let (result, exit_code) = wasi_exit_code(result);
 
+        let pid = env.data(&store).pid();
+        let tid = env.data(&store).tid();
         tracing::trace!(
-            "wasi[{}:{}]::main exit (code = {:?})",
-            env.data(store).pid(),
-            env.data(store).tid(),
-            res
+            %pid,
+            %tid,
+            %exit_code,
+            error=result.as_ref().err().map(|e| e as &dyn std::error::Error),
+            "main exit",
         );
 
-        let exit_code = match &res {
-            Ok(_) => Errno::Success.into(),
-            Err(err) => match err.as_exit_code() {
-                Some(code) if code.is_success() => {
-                    // This is actually not an error, so we need to fix up the
-                    // result
-                    res = Ok(());
-                    Errno::Success.into()
-                }
-                Some(other) => other,
-                None => Errno::Noexec.into(),
-            },
-        };
-
         env.cleanup(store, Some(exit_code));
 
-        res
+        result
+    }
+
+    /// Start the WASI executable with async threads enabled.
+    #[allow(clippy::result_large_err)]
+    pub fn run_with_store_async(
+        self,
+        module: Module,
+        mut store: Store,
+    ) -> Result<(), WasiRuntimeError> {
+        let (_, env) = self.instantiate(module, &mut store)?;
+
+        env.data(&store).thread.set_status_running();
+
+        let tasks = env.data(&store).tasks().clone();
+        let pid = env.data(&store).pid();
+        let tid = env.data(&store).tid();
+
+        // The return value is passed synchronously and will block until the result
+        // is returned this is because the main thread can go into a deep sleep and
+        // exit the dedicated thread
+        let (tx, rx) = std::sync::mpsc::channel();
+
+        tasks.task_dedicated(Box::new(move || {
+            run_with_deep_sleep(store, None, env, tx);
+        }))?;
+
+        let result = rx.recv().expect(
+            "main thread terminated without a result, this normally means a panic occurred",
+        );
+        let (result, exit_code) = wasi_exit_code(result);
+
+        tracing::trace!(
+            %pid,
+            %tid,
+            %exit_code,
+            error=result.as_ref().err().map(|e| e as &dyn std::error::Error),
+            "main exit",
+        );
+
+        result
     }
 }
 
+/// Extract the exit code from a `Result<(), WasiRuntimeError>`.
+///
+/// We need this because calling `exit(0)` inside a WASI program technically
+/// triggers [`WasiError`] with an exit code of `0`, but the end user won't want
+/// that treated as an error.
+fn wasi_exit_code(
+    mut result: Result<(), WasiRuntimeError>,
```

Comment:
There already is a function for this on `WasiRuntimeError`: https://github.com/wasmerio/wasmer/blob/eec73b70f45e15e48be592009b12a692e4c5b4c1/lib/wasi/src/lib.rs#L194